File: 1905-count-sub-islands/1905-count-sub-islands.cpp
class Solution {
public:
    int directions[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}} ;
    void dfs(int i, int j, int n, int m, vector<vector<int>>& grid2) {
        if (i >= n || i < 0 || j < 0 || j >= m)
            return;

        grid2[i][j] = 0;
        // if ((i - 1) >= 0 && grid2[i - 1][j] == 1) {
        //     dfs(i - 1, j, n, m, grid2);
        // }
        // if ((j - 1) >= 0 && grid2[i][j - 1] == 1) {
        //     dfs(i, j - 1, n, m, grid2);
        // }
        // if ((i + 1) < n && grid2[i + 1][j] == 1) {
        //     dfs(i + 1, j, n, m, grid2);
        // }
        // if ((j + 1) < m && grid2[i][j + 1] == 1) {
        //     dfs(i, j + 1, n, m, grid2);
        // }
        for (auto& direction : directions) {
            int nrow = i + direction[0];
            int ncol = j + direction[1];
            if (nrow >= 0 && nrow < n && ncol >= 0 && ncol < m &&
                grid2[nrow][ncol] == 1)
                dfs(nrow, ncol, n, m, grid2);
        }
    }
    int countSubIslands(vector<vector<int>>& grid1,
                        vector<vector<int>>& grid2) {
        int count = 0;
        int n = grid1.size();
        int m = grid1[0].size();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid1[i][j] == 0 && grid2[i][j] == 1) {
                    dfs(i, j, n, m, grid2);
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid2[i][j] == 1) {
                    dfs(i, j, n, m, grid2);
                    count++;
                }
            }
        }
        return count;
    }
};

File: 1905-count-sub-islands/1905-count-sub-islands.java
class Solution {
public int countSubIslands(int[][] grid1, int[][] grid2) {
int n = grid1.length, m = grid1[0].length;
int[] drow = {-1, 0, 1, 0}, dcol = {0, 1, 0, -1};
int[][] vis = new int[n][m];
int ans = 0;
for(int i=0;i<n;i++) {
for(int j=0;j<m;j++) {
if(vis[i][j] == 0 && grid2[i][j] == 1) {
if(helper(i, j, n, m, drow, dcol, vis, grid1, grid2)) ans++;
}
}
}
return ans;
}
static boolean helper(int row, int col, int n, int m, int[] drow, int[] dcol, int[][] vis, int[][] grid1, int[][] grid2) {
vis[row][col] = 1;
        boolean isIsland=true;

if(grid1[row][col] == 0) isIsland= false;
for(int i=0;i<4;i++) {
int nrow = row + drow[i];
int ncol = col + dcol[i];
if(nrow >= 0 && nrow < n && ncol >= 0 && ncol < m && vis[nrow][ncol] == 0 && grid2[nrow][ncol] == 1) {
boolean isIsland1=helper(nrow, ncol, n, m, drow, dcol, vis, grid1, grid2);
    isIsland=isIsland&&isIsland1;
}
}
return isIsland;
}
}